File: api/app/models/post.py
from sqlalchemy import (
    Column,
    Integer,
    String,
    DateTime,
    ForeignKey
)

from sqlalchemy.sql import func

from sqlalchemy.orm import relationship

from app.db.base import Base


class Post(Base):

    __tablename__ = "posts"

    id = Column(
        Integer,
        primary_key=True,
        index=True
    )

    image_url = Column(
        String,
        nullable=False
    )

    caption = Column(
        String,
        nullable=True
    )

    user_id = Column(
        Integer,
        ForeignKey("users.id"),
        nullable=False
    )

    created_at = Column(
        DateTime(timezone=True),
        server_default=func.now()
    )

    user = relationship(
        "User",
        back_populates="posts"
    )

    comments = relationship(
        "Comment",
        back_populates="post",
        cascade="all, delete"
    )
    
    likes = relationship(
        "Like",
        back_populates="post",
        cascade="all, delete-orphan"
    )

File: web/types/post.ts
export type UserPost = {

  id: number

  image_url: string

  caption: string | null
}


export type UserPostsResponse = {

  posts: UserPost[]

  next_cursor: number | null

  has_more: boolean
}

export type PostDetail = {

  id: number

  image_url: string

  caption: string | null

  created_at: string

  user: {
    id: number
    username: string
    profile_image_url: string | null
  }

  comments: {
    id: number
    content: string
  }[]
}